components/sup/src/event/types.rs

```diff
@@ -90,23 +90,19 @@ pub trait EventMessage: Message + Sized {
     }
 }
 
-// TODO (CM): these are repetitive, and will only get more so as we
-// add events. Consider implementing via macro instead.
-
-impl EventMessage for ServiceStartedEvent {
-    fn event_metadata(&mut self, event_metadata: EventMetadata) {
-        self.event_metadata = Some(event_metadata);
-    }
-}
-
-impl EventMessage for ServiceStoppedEvent {
-    fn event_metadata(&mut self, event_metadata: EventMetadata) {
-        self.event_metadata = Some(event_metadata);
-    }
+macro_rules! event_msg_impl {
+    ($($t:ty)*) => {$(
+        impl EventMessage for $t {
+            fn event_metadata(&mut self, event_metadata: EventMetadata) {
+                self.event_metadata = Some(event_metadata);
+            }
+        }
+    )*};
 }
 
-impl EventMessage for HealthCheckEvent {
-    fn event_metadata(&mut self, event_metadata: EventMetadata) {
-        self.event_metadata = Some(event_metadata);
-    }
-}
+event_msg_impl! (
+    ServiceStartedEvent
+    ServiceStoppedEvent
+    ServiceUpdateStartedEvent
+    HealthCheckEvent
+);
```

Comment:
I think my preference here would be for the macro to take a single type, rather than multiple.

It'd be a bit more typing to invoke them, but it also means the macro itself has a single responsibility (just adding the code, rather than adding code and also iterating). It simplifies the macro (very slightly), but I tend to favor the simplest macro possible. It's also more apparent from a glance at the code below exactly what is going on (you don't need to ask what it needs all the types for). 

(As an macro implementation aside, it's also a little odd to _not_ have commas between the types, since they're basically being treated as arguments here.)

```rust
event_msg_impl!(ServiceStartedEvent);
event_msg_impl!(ServiceStoppedEvent);
event_msg_impl!(ServiceUpdateStartedEvent);
event_msg_impl!(HealthCheckEvent);
```

I think this particular example is arguably in a gray area, though, so I'm open to other thoughts. @baumanj @raskchanky, what do you think?

Reply:
👍

Reply:
I'm fine with either approach, but if it takes multiple types, I agree that comma separation would be good (and should be straightforward to add).

Focusing more on this code, I think it's kind of confusing that the method is named `event_metadata`, but also assigns to a field named `event_metadata`. I'd expect the setter method to have a name like `set_metadata` instead.

Thinking about it a bit more, if `EventMetadata` is part of every event,  could we restructure a bit and avoid the need to have it as part of the message types at all? It's only ever set in the `publish` method, and since the argument to `publish` is moved, it can't be accessed afterward. Couldn't `publish` itself create and send the `EventMetadata` bytes and then the particular message type (which no longer has `EventMetadata` embedded)? If we can go this route, then we wouldn't need the macro at all.

Reply:
@baumanj I agree that the `EventMetadata` bit is a little weird. However, `publish` is using the `to_bytes` method from `EventMessage` to transform the entire event type into a protobuf byte string. Splitting that byte generation up would probably be pretty error prone, and more complex than the current (admittedly weird) code.

I'm open to other alternatives, though. This is probably the part of this code that I like the least.

Reply:
Sending the event on the wire together with its metadata is fine, but I think there must be some way to have `publish` handle the population of the bits that are identical across all types. Possibly it would require more types in the .proto, but I think that's preferable to the shenanigans we've got currently.